drivers/i2c/i2c_mchp_xec.c

```diff
@@ -216,37 +254,84 @@ static int wait_completion(const struct device *dev)
 	int counter = 0;
 	uint32_t ba = config->base_addr;
 
-	/* Wait for transaction to be completed */
-	while (MCHP_I2C_SMB_STS_RO(ba) & MCHP_I2C_SMB_STS_PIN) {
-		ret = xec_spin_yield(&counter);
+	while (1) {
+		uint8_t status = MCHP_I2C_SMB_STS_RO(ba);
 
-		if (ret < 0) {
-			if (MCHP_I2C_SMB_STS_RO(ba) & MCHP_I2C_SMB_STS_PIN) {
-				recover_from_error(dev);
-				return ret;
-			}
+		/* Is bus error ? */
+		if (status & MCHP_I2C_SMB_STS_BER) {
+			recover_from_error(dev);
+			return -EBUSY;
 		}
-	}
 
-	/* Check if Slave send ACK/NACK */
-	if (MCHP_I2C_SMB_STS_RO(ba) & MCHP_I2C_SMB_STS_LRB_AD0) {
-		recover_from_error(dev);
-		return -EIO;
-	}
+		/* Is Lost arbitration ? */
+		status = MCHP_I2C_SMB_STS_RO(ba);
+		if (status & MCHP_I2C_SMB_STS_LAB) {
+			return -EPERM;
+		}
 
-	/* Check for bus error */
-	if (MCHP_I2C_SMB_STS_RO(ba) & MCHP_I2C_SMB_STS_BER) {
-		recover_from_error(dev);
-		return -EBUSY;
+		status = MCHP_I2C_SMB_STS_RO(ba);
+		/* PIN -> 0 indicates I2C is done */
+		if (!(status & MCHP_I2C_SMB_STS_PIN)) {
+			/* PIN == 0. LRB contains state of 9th bit */
+			if (status & MCHP_I2C_SMB_STS_LRB_AD0) { /* NACK? */
+				/* Send STOP */
+				MCHP_I2C_SMB_CTRL_WO(ba) =
+						MCHP_I2C_SMB_CTRL_PIN |
+						MCHP_I2C_SMB_CTRL_ESO |
+						MCHP_I2C_SMB_CTRL_STO |
+						MCHP_I2C_SMB_CTRL_ACK;
+				k_busy_wait(BUS_IDLE_US_DFLT);
+				return -EIO;
+			}
+			break; /* success: ACK */
+		}
+
+		ret = xec_spin_yield(&counter);
+		if (ret < 0) {
+			return ret;
+		}
 	}
 
 	return 0;
 }
 
-static bool check_lines(uint32_t ba)
+/*
+ * Call GPIO driver to read state of pins.
+ * Return boolean true if both lines HIGH else return boolean false
+ */
+static bool check_lines_high(const struct device *dev)
 {
-	return ((!(MCHP_I2C_SMB_BB_CTRL(ba) & MCHP_I2C_SMB_BB_CLKI_RO)) ||
-		(!(MCHP_I2C_SMB_BB_CTRL(ba) & MCHP_I2C_SMB_BB_DATI_RO)));
+	const struct i2c_xec_config *config =
+		(const struct i2c_xec_config *const)(dev->config);
+	struct i2c_xec_data *data = (struct i2c_xec_data *const)(dev->data);
+	uint32_t i2c_lines = 0U;
+	gpio_port_value_t sda = 0, scl = 0;
+
+	if (gpio_port_get_raw(data->sda_gpio, &sda)) {
+		LOG_DBG("gpio_port_get_raw for %s SDA failed", dev->name);
+		return false;
+	}
+
+	if (sda & BIT(config->sda_pos)) {
+		i2c_lines |= I2C_LINES_SDA_HI;
+	}
+
+	/* both pins on same GPIO port? */
```

Comment:
This seems like it'd be a more serious issue, having scl and sda on the same gpio? That's not really i2c anymore then I would think. Assuming they are always on different pins here could really simplify this function, perhaps something like...

```c
int ret;

ret  = gpio_port_raw_get(..., &sda);
ret |= gpio_port_raw_get(..., &scl);
if (ret) {
   LOG_ERR("failed to get gpio ...");
   return false;
}
return (sda & BIT(config->sda_pos)) && (scl & BIT(config-> scl_pos));
```

Reply:
Simplified the function.